From 089b2e7065b73cebb1f75099997d8f0c2194759d Mon Sep 17 00:00:00 2001
From: Mika Lindqvist <postmaster@raasu.org>
Date: Sat, 19 Jun 2021 00:10:44 +0300
Subject: [PATCH 10/13] [chunkset_neon] Use vdupq_n_u64. * Using vdupq_n_u64
 duplicates the unsigned 64-bit integer to two consecutive aligned memory
 locations in stack so compiler can use wider load instructions.   All
 different-sized general-purpose registers overlay on ARM/AArch64, so any
 vector cast is no-op in assembly.

---
 arch/arm/chunkset_neon.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/arm/chunkset_neon.c b/arch/arm/chunkset_neon.c
index b1fcb24..e0ad3e0 100644
--- a/arch/arm/chunkset_neon.c
+++ b/arch/arm/chunkset_neon.c
@@ -37,7 +37,9 @@ static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
 }
 
 static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
-    *chunk = vcombine_u8(vld1_u8(from), vld1_u8(from));
+    uint64_t tmp;
+    memcpy(&tmp, from, 8);
+    *chunk = vreinterpretq_u8_u64(vdupq_n_u64(tmp));
 }
 
 #define CHUNKSIZE        chunksize_neon
-- 
2.32.0

